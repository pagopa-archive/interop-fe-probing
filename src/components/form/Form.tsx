import { Grid, Button, FormHelperText, Card, Typography, TextField } from '@mui/material'
import { Controller, useForm } from 'react-hook-form'

interface Field {
  name: string
  label: string
  type: string
  rules: { [key: string]: any }
  width?: string
}

interface FormData {
  fields: Array<Field>
  submitButton: string
  submitFunction: Function
  defaultValues: { [key: string]: string }
}

export const Form: React.FC<FormData> = ({
  fields,
  submitButton,
  submitFunction,
  defaultValues,
}) => {
  /**
   * form functionalities from react-hook-forms
   */
  const {
    handleSubmit,
    control,
    formState: { errors, isValid },
  } = useForm({
    defaultValues: defaultValues,
    mode: 'onSubmit',
    reValidateMode: 'onSubmit',
  })

  return (
    <Card
      elevation={16}
      sx={{
        p: 4,
        borderRadius: '16px',
        maxWidth: '480px',
      }}
    >
      <form onSubmit={handleSubmit((data) => submitFunction(data))}>
        <Grid item container direction="column" rowSpacing={3}>
          <Grid item container rowSpacing={3}>
            {fields.map((field: Field) => (
              <Grid item container key={field.name}>
                <Controller
                  control={control}
                  name={field.name}
                  rules={field.rules}
                  render={({
                    field: { onChange, onBlur, value, name, ref },
                    fieldState: { invalid, isTouched, isDirty, error },
                    formState,
                  }) => (
                    <>
                      <TextField
                        type={field.type}
                        label={field.label}
                        value={value}
                        onChange={onChange}
                        sx={{
                          ...(field.width && {
                            width: field.width,
                          }),
                        }}
                      />
                      <FormHelperText error>
                        {errors[field.name] ? errors[field.name]?.message : ' '}
                      </FormHelperText>
                    </>
                  )}
                />
              </Grid>
            ))}
          </Grid>

          <Grid item alignSelf="center">
            <Button
              sx={{
                backgroundColor: 'primary.main',
                '&:hover': { backgroundColor: 'primary.dark' },
              }}
              size="large"
              type="submit"
              variant="contained"
              disabled={!isValid}
            >
              <Typography sx={{ color: 'white' }}> {submitButton}</Typography>
            </Button>
          </Grid>
        </Grid>
      </form>
    </Card>
  )
}
