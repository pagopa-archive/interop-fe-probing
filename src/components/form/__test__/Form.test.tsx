import { render } from '@testing-library/react'
import { Form } from '../Form'
import { expect, test, vi } from 'vitest'

const fields = [
  {
    name: 'username',
    label: 'username',
    type: 'email',
    rules: {
      required: 'username required',
    },
  },
  {
    name: 'password',
    label: 'password',
    type: 'password',
    rules: {
      required: 'password required',
    },
  },
]

const defaultFormValues: { [key: string]: string } = {
  username: '',
  password: '',
}

const onSubmit = vi.fn()

describe('Form', () => {
  test('render component', () => {
    const { container } = render(
      <Form
        fields={fields}
        submitButton={'Submit'}
        submitFunction={onSubmit}
        defaultValues={defaultFormValues}
        buttonDisable={'onValid'}
        validationMode={'onSubmit'}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
