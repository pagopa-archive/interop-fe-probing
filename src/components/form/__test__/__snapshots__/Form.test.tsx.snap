// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form > render component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation16 MuiCard-root css-9cvwhb-MuiPaper-root-MuiCard-root"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column css-ljy4ha-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item css-1qy0967-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="username"
                id="username-label"
              >
                username
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="username"
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      username
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error css-1d1r5q-MuiFormHelperText-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                password
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="password"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error css-1d1r5q-MuiFormHelperText-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-131xi47-MuiGrid-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-w4y956-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-jb1ygz-MuiTypography-root"
            >
               
              Submit
            </p>
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
